### PR TITLE
✨ #538: Add refresh tokens

### PR DESCRIPTION
**Pull request by samrichca**
- Refresh the user's JWT on page change and when fetching from DACO-API.

### CODE REVIEW

File: global/utils/auth/refreshJwt.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import { isValidJwt } from '../egoTokenUtils';
+
+const { NEXT_PUBLIC_EGO_API_ROOT, NEXT_PUBLIC_EGO_CLIENT_ID } = getConfig();
+const egoRefreshUrl = urlJoin(
+  NEXT_PUBLIC_EGO_API_ROOT,
+  `/oauth/refresh?client_id=${NEXT_PUBLIC_EGO_CLIENT_ID}`,
+);
+
+var maxConcurrent = 1;
+var maxQueue = Infinity;
```
Comment (ciaranschutte) [inline global/utils/auth/refreshJwt.tsx:33]: I haven't looked too closely at promise-queue module, but do we need these on global scope?

Comment (samrichca) [reply global/utils/auth/refreshJwt.tsx:33]: That is a really good question, it works this way on platform-UI so I didn't question it. I'll look into it.

Comment (anncatton) [reply global/utils/auth/refreshJwt.tsx:33]: i think that keeps us from having several instances of queue?

Comment (ciaranschutte) [reply global/utils/auth/refreshJwt.tsx:33]: alternative is a singleton I guess? It's a pretty old module. most of the commits past 2015 are just dependency updates and versioning so wondering if var usage is needed or just older lib.  if it works, it works :)

File: pages/_app.tsx [hunk 2]
```diff
@@ -64,22 +65,43 @@ const App = ({
...
-      // redirect to logout when token is expired/missing only if user is on a non-public page
-      if (!Component.isPublic) {
-        Router.push({
-          pathname: '/',
-          query: { session_expired: true },
-        });
+    const handleAuth = async () => {
+      console.log('APP handleAuth');
+      const egoJwt = localStorage.getItem(EGO_JWT_KEY) || '';
+      if (isValidJwt(egoJwt)) {
```
[the remark this answers is not in the document]
Comment (anncatton) [reply pages/_app.tsx:84]: sorry my indentation didn't translate here 😕

File: pages/_app.tsx [hunk 3]
```diff
@@ -64,22 +65,43 @@ const App = ({
     };
   }, []);
 
+  const logout = () => {
```
Comment (blabadi) [inline pages/_app.tsx:68]: should the logout call DELETE /oauth/refresh ?

Comment (anncatton) [reply pages/_app.tsx:68]: ah good catch!

File: pages/_app.tsx [hunk 4]
```diff
@@ -64,22 +65,43 @@ const App = ({
...
+    setInitialJwt('');
+    localStorage.removeItem(EGO_JWT_KEY);
+    // redirect to logout when token is expired/missing only if user is on a non-public page
+    if (!Component.isPublic) {
+      Router.push({
+        pathname: '/',
+        query: { session_expired: true },
```
[the remark this answers is not in the document]
Comment (samrichca) [reply pages/_app.tsx:75]: Good idea! Removed it in logout() in AuthContext, which is called by the nav bar logout button.

File: pages/_app.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (samrichca) [reply pages/_app.tsx]: I understood what you meant though - updated that part.

File: pages/_app.tsx (hunk 3 above)
Comment (samrichca) [reply pages/_app.tsx:72]: I added a delete call, and I'm deleting the localStorage JWT after the call is complete.

Comment (blabadi) [reply pages/_app.tsx:72]: looks good thanks !

**Pull request merged**